File: 2022/src/day16.rb
data = <<-STR
Valve AA has flow rate=0; tunnels lead to valves DD, II, BB
Valve BB has flow rate=13; tunnels lead to valves CC, AA
Valve CC has flow rate=2; tunnels lead to valves DD, BB
Valve DD has flow rate=20; tunnels lead to valves CC, AA, EE
Valve EE has flow rate=3; tunnels lead to valves FF, DD
Valve FF has flow rate=0; tunnels lead to valves EE, GG
Valve GG has flow rate=0; tunnels lead to valves FF, HH
Valve HH has flow rate=22; tunnel leads to valve GG
Valve II has flow rate=0; tunnels lead to valves AA, JJ
Valve JJ has flow rate=21; tunnel leads to valve II
STR
# data = File.read('./tasks/day16.txt')

graph = data.lines.each_with_object({}) do |line, hash|
  valve, tunnels = line.split('; ')
  parts = valve.split
  name = parts[1]
  _, flow_rate = parts[4].split('=')
  tunnels = tunnels.split[4..].map { _1.gsub(',', '')}

  hash[name] = {name:, flow_rate: flow_rate.to_i, tunnels:}
end

CACHE = {}
def call_cached(graph, v, depth, rate, enabled)
  key = [v[:name], depth, rate]
  res = CACHE[key]
  return res if res

  res = dfs(graph, v, depth, rate, enabled)
  CACHE[key] = res
  res
end

def dfs(graph, v, depth=30, rate=0, enabled=[])
  return 0 if depth == 0

  results = v[:tunnels].map do |t|
    call_cached(graph, graph[t], depth - 1, rate, enabled) + rate
  end
  if !enabled.include?(v[:name]) && v[:flow_rate] > 0
    results << (call_cached(graph, v, depth - 1, rate + v[:flow_rate], [*enabled, v[:name]]) + rate)
  end

  results.max
end

pp dfs(graph, graph['AA'])

# part 1 = 1986

CACHE_2 = {}
def call_cached_2(graph, v, w, depth, rate, enabled)
  key = [v[:name], w[:name], depth, rate, enabled.hash]
  res = CACHE_2[key]
  return res if res

  res1 = dfs_2(graph, v, w, depth, rate, enabled)
  # if res && res != res1
    # pp(v: v[:name], w: w[:name], depth:, rate:, enabled:)
    # pp(key:, res:, res1:)

    # exit 1
  # end

  CACHE_2[key] = res1
  res1
end

def dfs_2(graph, v, w, depth=26, rate=0, enabled=[])
  return 0 if depth == 0

  results = v[:tunnels].product(w[:tunnels]).map do |t, el|
    call_cached_2(graph, graph[t], graph[el], depth - 1, rate, enabled) + rate
  end
  if !enabled.include?(v[:name]) && v[:flow_rate] > 0 && !enabled.include?(w[:name]) && w[:flow_rate] > 0 && v[:name] != w[:name]
    results << (call_cached_2(graph, v, w, depth - 1, rate + v[:flow_rate] + w[:flow_rate], [*enabled, v[:name], w[:name]]) + rate)
  end
  if !enabled.include?(v[:name]) && v[:flow_rate] > 0
    w[:tunnels].each do |el|
      results << (call_cached_2(graph, v, graph[el], depth - 1, rate + v[:flow_rate], [*enabled, v[:name]]) + rate)
    end
  end
  if !enabled.include?(w[:name]) && w[:flow_rate] > 0
    v[:tunnels].each do |t|
      results << (call_cached_2(graph, graph[t], w, depth - 1, rate + w[:flow_rate], [*enabled, w[:name]]) + rate)
    end
  end

  results.max
end

pp dfs_2(graph, graph['AA'], graph['AA'])

# part 2 = 2431 - too low

File: 2022/src/day16.rs
use std::{collections::{HashMap, hash_map::DefaultHasher}, fmt::Display};
use std::hash::{Hash, Hasher};

#[derive(Debug)]
struct Valve {
    name: String,
    rate: u32,
    tunnels: Vec<String>,
}

impl From<&str> for Valve {
    fn from(string: &str) -> Self {
        let (valve, tunnels) = string.split_once(";").unwrap();
        let mut parts = valve.split_whitespace();
        parts.next();
        let name = parts.next().unwrap().to_string();
        let rate = parts.skip(2).next().unwrap();
        let rate = rate.split("=").skip(1).next().unwrap().parse().unwrap();
        let tunnels: Vec<_> = tunnels
            .split_whitespace()
            .skip(4)
            .map(|s| s.trim_matches(',').to_string())
            .collect();

        Valve {
            name,
            rate,
            tunnels,
        }
    }
}

pub fn solve(input: &str) -> (Box<dyn Display>, Box<dyn Display>) {
    (
        Box::new(solve_first_part(input)),
        Box::new(solve_second_part(input)),
    )
}

fn solve_first_part(input: &str) -> u32 {
    // let valves: HashMap<_, _> = input
    //     .lines()
    //     .map(|l| {
    //         let valve = Valve::from(l);
    //         (valve.name.clone(), valve)
    //     })
    //     .collect();

    // dbg!(&valves);

    // dfs(&valves, valves.get("AA").unwrap(), 30, 0, Vec::new())
    0
}

fn solve_second_part(input: &str) -> i32 {
    42
}

// fn dfs(
//     graph: &HashMap<String, Valve>,
//     v: &Valve,
//     depth: u32,
//     rate: u32,
//     enabled: Vec<String>,
// ) -> u32 {
//     let cache = HashMap::new();

//     let inner = |graph, v: &Valve, depth, rate, enabled: Vec<String>| {
//         // let key =
//         // if depth == 0 {
//         //     return 0;
//         // }

//         let mut results: Vec<u32> = v
//             .tunnels
//             .iter()
//             .map(|t| {
//                 dfs(
//                     graph,
//                     graph.get(t).unwrap(),
//                     depth - 1,
//                     rate,
//                     enabled.clone(),
//                 ) + rate
//             })
//             .collect();
//         if enabled.contains(&v.name) && v.rate > 0 {
//             let mut enabled = enabled.clone();
//             enabled.push(v.name.clone());
//             results.push(dfs(graph, v, depth - 1, rate + v.rate, enabled))
//         }

//         *results.iter().max().unwrap()
//     };

//     inner(graph, v, depth, rate, enabled)
// }

// fn calculate_hash<T: Hash>(t: &T) -> u64 {
//     let mut s = DefaultHasher::new();
//     t.hash(&mut s);
//     s.finish()
// }

#[cfg(test)]
mod tests {
    use std::fs;

    use super::*;

    const INPUT: &str = "Valve AA has flow rate=0; tunnels lead to valves DD, II, BB
Valve BB has flow rate=13; tunnels lead to valves CC, AA
Valve CC has flow rate=2; tunnels lead to valves DD, BB
Valve DD has flow rate=20; tunnels lead to valves CC, AA, EE
Valve EE has flow rate=3; tunnels lead to valves FF, DD
Valve FF has flow rate=0; tunnels lead to valves EE, GG
Valve GG has flow rate=0; tunnels lead to valves FF, HH
Valve HH has flow rate=22; tunnel leads to valve GG
Valve II has flow rate=0; tunnels lead to valves AA, JJ
Valve JJ has flow rate=21; tunnel leads to valve II";

    #[ignore]
    #[test]
    fn test_first_part() {
        let answer = 1651;

        assert_eq!(answer, solve_first_part(INPUT))
    }

    #[ignore]
    #[test]
    fn test_second_part() {
        let answer = 1707;

        assert_eq!(answer, solve_second_part(INPUT))
    }

    // check_answers!(42, 42);
}
